build_tools/autogen_ltc_backend.py

```diff
@@ -0,0 +1,290 @@
+        yaml.dump(
+            {
+                "backend": "Lazy",
+                "cpp_namespace": "torch_lazy_tensors",
+                "full_codegen": opnames,
+                "supported": sorted(supported_ops),
+            },
+            f,
+            default_flow_style=False,
+        )
+        f.write(
+            dedent(
+                """
+
+                # Skipped ops (supported by Torch-MLIR but no equivalent native function)
+                """
+            )
+            + os.linesep.join(f"#  - {op}" for op in sorted(skipped))
+        )
+
+    return parsed_yaml, grouped_native_functions
+
+
+@dataclass(frozen=True)
+class MlirLazyIr(codegen.gen_lazy_tensor.dest.LazyIR):
+    lowering_function_type: str = "torch::lazy::MlirFunction"
+    lowering_context_type: str = "torch::lazy::MlirLoweringContext*"
+    lowering_return_type: str = "torch::lazy::MlirOpVector"
+
+    def lowering_body(self, f):
+        func = (
+            f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        )
+        schema = LazyIrSchema(func)
+
+        return f"""
+        UNIMPLEMENTED_ERROR(
+            "'{func}' lowering not yet implemented"
+        );
+    """.rstrip()
+
+
+def generate_backend(
+    source_yaml: Path, backend_path: Path, parsed_yaml: dict, grouped_native_functions: list
+):
+    print("Running Lazy Tensor Autogen")
+
+    # No fallback code allowed
+    def gen_fallback_code(*args, **kwargs):
+        return ""
+
+    codegen.dest.lazy_ir.gen_fallback_code = gen_fallback_code
+
+    codegen.gen_lazy_tensor.run(
+        source_yaml=str(source_yaml),
+        output_dir=str(backend_path),
+        dry_run=False,
+        impl_path=str(backend_path.joinpath("aten_ltc_mlir_type.cpp")),
+        gen_ts_lowerings=False,
+        node_base="torch::lazy::MlirNode",
+        node_base_hdr=str(backend_path.joinpath("mlir_node.h")),
+        tensor_class="torch::lazy::LazyTensor",
+        tensor_class_hdr="torch/csrc/lazy/core/tensor.h",
+        lazy_ir_cls=MlirLazyIr,
+    )
+
+    # Remove lazy_tensor_core imports
+    subprocess.check_call(
+        [
+            "sed",
+            "-i",
+            "/lazy_tensor_core/d",
+            str(backend_path.joinpath("LazyNativeFunctions.cpp")),
+        ]
+    )
+
+    # Autogenerate shape inference placeholders
+    import re
+
+    sig_re = re.compile(f"std::vector<Shape> (?P<name>[_a-zA-Z0-9]+)\((?P<signature>.+)\);")
+    shape_inference_decls = backend_path.joinpath("LazyShapeInference.h").read_text()
+
+    shape_inference_defs = []
+    for name, signature in sig_re.findall(shape_inference_decls):
+        shape_inference_defs.append(
+            dedent(
+                f"""
+                std::vector<Shape> {name}({signature}) {{
+                    UNIMPLEMENTED_ERROR("{name}");
+                }}
+                """
+            )
+        )
+
+    backend_path.joinpath("LazyShapeInference.cpp").write_text(
```

Comment:
note that I've just landed https://github.com/pytorch/pytorch/pull/72756 so you may need to update this part.

Reply:
e.g. you probably shouldn't have to do anything at all for shape inference, other than contribute new functions to it if found missing.

Reply:
@wconstab are lazy shape inference declarations not being autogenerated at all anymore?

Reply:
correct: the new way is the header and impl is handwritten, but the generator spits out a message saying 'missing this function please add it' and gives you a signature you can copy/paste into the header.  This way backends can share the inference functions and there isn't an issue that each backend generates a different header that compiles against a shared implementation.

Reply:
Got it. Thanks!